File: web/e-platform-web/src/services/StudentService.js
const API_URL = process.env.REACT_APP_BFF_API_URL || 'https://tunsiaed.onrender.com';
const getAuthHeaders = (token) => ({
  'Authorization': `Bearer ${token}`,
  'Content-Type': 'application/json',
});

class StudentService {
  // ====================================================================
  // DIRECT SERVICE ENDPOINTS (single module operations)
  // These call the Student module directly without orchestration
  // ====================================================================

  /**
   * Fetches student dashboard statistics
   * DIRECT: Single module operation
   * @param {string} token - Authentication token
   * @returns {Promise<Object>} Student stats data
   */
  static async getStats(token) {
    const res = await fetch(`${API_URL}/api/v1/student/stats`, { headers: getAuthHeaders(token) });
    if (!res.ok) throw new Error('Failed to fetch student stats');
    return res.json();
  }

  /**
   * Fetches courses available to the student
   * DIRECT: Single module operation
   * @param {string} token - Authentication token
   * @returns {Promise<Array>} List of available courses
   */
  static async getCourses(token) {
    const res = await fetch(`${API_URL}/api/v1/student/courses`, { headers: getAuthHeaders(token) });
    if (!res.ok) throw new Error('Failed to fetch student courses');
    return res.json();
  }

  /**
   * Fetches student's enrollments
   * DIRECT: Single module operation
   * @param {string} token - Authentication token
   * @returns {Promise<Array>} List of student's enrollments
   */
  static async getEnrollments(token) {
    const res = await fetch(`${API_URL}/api/v1/student/enrollments`, { headers: getAuthHeaders(token) });
    if (!res.ok) throw new Error('Failed to fetch student enrollments');
    return res.json();
  }

  /**
   * Fetches student's certificates
   * DIRECT: Single module operation
   * @param {string} token - Authentication token
   * @returns {Promise<Array>} List of certificates
   */
  static async getCertificates(token) {
    const res = await fetch(`${API_URL}/api/v1/student/certificates`, { headers: getAuthHeaders(token) });
    if (!res.ok) throw new Error('Failed to fetch student certificates');
    return res.json();
  }

  /**
   * Fetches student's learning progress (basic)
   * DIRECT: Single module operation
   * @param {string} token - Authentication token
   * @returns {Promise<Object>} Progress data
   */
  static async getProgress(token) {
    const res = await fetch(`${API_URL}/api/v1/student/progress`, { headers: getAuthHeaders(token) });
    if (!res.ok) throw new Error('Failed to fetch student progress');
    return res.json();
  }

  /**
   * Updates student's progress (basic)
   * DIRECT: Single module operation
   * @param {string} token - Authentication token
   * @param {Object} progressData - Progress data to update
   * @returns {Promise<Object>} Updated progress
   */
  static async updateProgress(token, progressData) {
    const res = await fetch(`${API_URL}/api/v1/student/progress`, {
      method: 'PATCH',
      headers: getAuthHeaders(token),
      body: JSON.stringify(progressData),
    });
    if (!res.ok) throw new Error('Failed to update student progress');
    return res.json();
  }

  // ====================================================================
  // ORCHESTRATED ENDPOINTS (cross-module operations)
  // These use StudentDashboardOrchestrator for aggregated data
  // ====================================================================

  /**
   * Fetches complete student dashboard data (aggregated)
   * ORCHESTRATOR: Aggregates stats, courses, enrollments, certificates, progress
   * Cross-module: Student + Progress + Enrollment + Certificate modules
   * @param {string} token - Authentication token
   * @returns {Promise<Object>} { stats, courses, enrollments, certificates, progress }
   */
  static async getDashboard(token) {
    const res = await fetch(`${API_URL}/api/v1/student/dashboard`, { 
      headers: getAuthHeaders(token) 
    });
    if (!res.ok) {
      const errorData = await res.json().catch(() => ({}));
      throw new Error(errorData.error || 'Failed to fetch student dashboard');
    }
    return res.json();
  }

  /**
   * Fetches enrollments with detailed progress data
   * ORCHESTRATOR: Merges enrollment and progress data
   * Cross-module: Student + Progress modules
   * @param {string} token - Authentication token
   * @returns {Promise<Array>} Enrollments with progress details
   */
  static async getEnrollmentsWithProgress(token) {
    const res = await fetch(`${API_URL}/api/v1/student/enrollments/detailed`, { 
      headers: getAuthHeaders(token) 
    });
    if (!res.ok) {
      const errorData = await res.json().catch(() => ({}));
      throw new Error(errorData.error || 'Failed to fetch enrollments with progress');
    }
    return res.json();
  }

  /**
   * Fetches comprehensive learning overview
   * ORCHESTRATOR: Combines stats, progress, certificates for complete view
   * Cross-module: Student + Progress + Certificate modules
   * @param {string} token - Authentication token
   * @returns {Promise<Object>} { stats, progressSummary, recentCertificates, learningStreak }
   */
  static async getLearningOverview(token) {
    const res = await fetch(`${API_URL}/api/v1/student/learning/overview`, { 
      headers: getAuthHeaders(token) 
    });
    if (!res.ok) {
      const errorData = await res.json().catch(() => ({}));
      throw new Error(errorData.error || 'Failed to fetch learning overview');
    }
    return res.json();
  }

  /**
   * Updates progress with enrollment synchronization
   * ORCHESTRATOR: Updates progress and keeps enrollment in sync
   * Cross-module: Student + Progress + Enrollment modules
   * @param {string} token - Authentication token
   * @param {Object} progressData - { progressId, updateData }
   * @returns {Promise<Object>} Updated progress with enrollment status
   */
  static async updateProgressOrchestrated(token, progressData) {
    const res = await fetch(`${API_URL}/api/v1/student/progress/update`, {
      method: 'POST',
      headers: getAuthHeaders(token),
      body: JSON.stringify(progressData),
    });
    if (!res.ok) {
      const errorData = await res.json().catch(() => ({}));
      throw new Error(errorData.error || 'Failed to update progress');
    }
    return res.json();
  }

  /**
   * Fetches progress by enrollment ID
   * ORCHESTRATOR: Gets detailed progress for specific enrollment
   * Cross-module: Student + Progress modules
   * @param {string} token - Authentication token
   * @param {string} enrollmentId - The enrollment ID
   * @returns {Promise<Object>} Detailed progress for enrollment
   */
  static async getProgressByEnrollment(token, enrollmentId) {
    const res = await fetch(`${API_URL}/api/v1/student/progress/enrollment/${enrollmentId}`, { 
      headers: getAuthHeaders(token) 
    });
    if (!res.ok) {
      const errorData = await res.json().catch(() => ({}));
      throw new Error(errorData.error || 'Failed to fetch enrollment progress');
    }
    return res.json();
  }

  /**
   * Fetches progress by course ID
   * ORCHESTRATOR: Gets course-specific progress summary
   * Cross-module: Student + Progress modules
   * @param {string} token - Authentication token
   * @param {string} courseId - The course ID
   * @returns {Promise<Object>} Progress summary for course
   */
  static async getProgressByCourse(token, courseId) {
    const res = await fetch(`${API_URL}/api/v1/student/progress/course/${courseId}`, { 
      headers: getAuthHeaders(token) 
    });
    if (!res.ok) {
      const errorData = await res.json().catch(() => ({}));
      throw new Error(errorData.error || 'Failed to fetch course progress');
    }
    return res.json();
  }

  /**
   * Fetches overall progress overview
   * ORCHESTRATOR: Gets all courses progress summary
   * Cross-module: Student + Progress modules
   * @param {string} token - Authentication token
   * @returns {Promise<Object>} Overall progress overview
   */
  static async getProgressOverview(token) {
    const res = await fetch(`${API_URL}/api/v1/student/progress/overview`, { 
      headers: getAuthHeaders(token) 
    });
    if (!res.ok) {
      const errorData = await res.json().catch(() => ({}));
      throw new Error(errorData.error || 'Failed to fetch progress overview');
    }
    return res.json();
  }
}

export default StudentService;

File: mobile/e_platform_mobile/lib/Service/student/student_service.dart
import '../../core/api/api_client.dart';
import '../../core/api/api_endpoints.dart';

/// Student Service
/// Handles all student-related API calls
class StudentService {
  final ApiClient _apiClient = ApiClient();

  /// Get student dashboard
  Future<Map<String, dynamic>> getStudentDashboard() async {
    final response = await _apiClient.get(
      ApiEndpoints.studentDashboard,
    );
    return response.data;
  }

  /// Get student courses (enrolled courses)
  Future<List<Map<String, dynamic>>> getStudentCourses() async {
    final response = await _apiClient.get(
      ApiEndpoints.studentCourses,
    );
    return List<Map<String, dynamic>>.from(response.data);
  }

  /// Get student progress
  Future<Map<String, dynamic>> getStudentProgress() async {
    final response = await _apiClient.get(
      ApiEndpoints.studentProgress,
    );
    return response.data;
  }

  /// Get student progress for specific course
  Future<Map<String, dynamic>> getCourseProgress(String courseId) async {
    final response = await _apiClient.get(
      ApiEndpoints.studentProgressByCourse(courseId),
    );
    return response.data;
  }

  /// Get student certificates
  Future<List<Map<String, dynamic>>> getStudentCertificates() async {
    final response = await _apiClient.get(
      ApiEndpoints.studentCertificates,
    );
    return List<Map<String, dynamic>>.from(response.data);
  }

  /// Get student transactions
  Future<List<Map<String, dynamic>>> getStudentTransactions() async {
    final response = await _apiClient.get(
      '${ApiEndpoints.transactions}/my-transactions',
    );
    return List<Map<String, dynamic>>.from(response.data);
  }

  // ========== ORCHESTRATOR METHODS (Cross-Module Operations) ==========

  /// Get enrollments with detailed progress (ORCHESTRATOR)
  /// Cross-module: Student + Progress modules
  /// Returns enrollments merged with progress details
  Future<Map<String, dynamic>> getEnrollmentsWithProgress() async {
    final response = await _apiClient.get(
      ApiEndpoints.studentEnrollmentsDetailed,
    );
    return response.data;
  }

  /// Get learning overview (ORCHESTRATOR)
  /// Cross-module: Student + Progress + Certificate modules
  /// Returns comprehensive learning stats, progress, and certificates
  Future<Map<String, dynamic>> getLearningOverview() async {
    final response = await _apiClient.get(
      ApiEndpoints.studentLearningOverview,
    );
    return response.data;
  }

  /// Get progress overview (ORCHESTRATOR)
  /// Cross-module: Student + Progress modules
  /// Returns summary of progress across all courses
  Future<Map<String, dynamic>> getProgressOverview() async {
    final response = await _apiClient.get(
      ApiEndpoints.studentProgressOverview,
    );
    return response.data;
  }

  /// Get progress by enrollment (ORCHESTRATOR)
  /// Cross-module: Progress + Enrollment modules
  /// Returns detailed progress for specific enrollment
  Future<Map<String, dynamic>> getProgressByEnrollment(String enrollmentId) async {
    final response = await _apiClient.get(
      ApiEndpoints.studentProgressByEnrollment(enrollmentId),
    );
    return response.data;
  }

  /// Get progress by course (ORCHESTRATOR)
  /// Cross-module: Progress + Enrollment + Course modules
  /// Returns course-specific progress summary
  Future<Map<String, dynamic>> getProgressByCourse(String courseId) async {
    final response = await _apiClient.get(
      ApiEndpoints.studentProgressByCourse(courseId),
    );
    return response.data;
  }

  /// Update progress with orchestration (ORCHESTRATOR)
  /// Cross-module: Progress + Enrollment modules
  /// Updates progress and syncs with enrollment automatically
  Future<Map<String, dynamic>> updateProgressOrchestrated(Map<String, dynamic> progressData) async {
    final response = await _apiClient.post(
      ApiEndpoints.studentProgressUpdate,
      data: progressData,
    );
    return response.data;
  }

  /// Get student stats (DIRECT)
  /// Single module operation
  Future<Map<String, dynamic>> getStudentStats() async {
    final response = await _apiClient.get(
      ApiEndpoints.studentStats,
    );
    return response.data;
  }

  /// Get student enrollments (DIRECT)
  /// Single module operation
  Future<List<Map<String, dynamic>>> getStudentEnrollments() async {
    final response = await _apiClient.get(
      ApiEndpoints.studentEnrollments,
    );
    return List<Map<String, dynamic>>.from(response.data);
  }

  /// Get student courses (DIRECT)
  /// Single module operation
  Future<List<Map<String, dynamic>>> getStudentCourses() async {
    final response = await _apiClient.get(
      ApiEndpoints.studentCourses,
    );
    return List<Map<String, dynamic>>.from(response.data);
  }

  /// Update student progress (DIRECT)
  /// Single module operation
  Future<Map<String, dynamic>> updateProgress(
    Map<String, dynamic> progressData,
  ) async {
    final response = await _apiClient.patch(
      ApiEndpoints.studentProgress,
      data: progressData,
    );
    return response.data;
  }
}
